pkg/generators/forwarding/fluentd/generators.go

```diff
@@ -0,0 +1,210 @@
+package fluentd
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+	"text/template"
+
+	logforward "github.com/openshift/cluster-logging-operator/pkg/apis/logging/v1alpha1"
+	"github.com/openshift/cluster-logging-operator/pkg/generators"
+	"github.com/openshift/cluster-logging-operator/pkg/logger"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+var (
+	defaultInfraProjectPatterns = []string{"default", "openshift", "openshift-", "kube-"}
+)
+
+//ConfigGenerator is a config generator for fluentd
+type ConfigGenerator struct {
+	*generators.Generator
+}
+
+//NewConfigGenerator creates an instance of FluentdConfigGenerator
+func NewConfigGenerator() (*ConfigGenerator, error) {
+	engine, err := generators.New("OutputLabelConf",
+		&template.FuncMap{
+			"labelName":           labelName,
+			"sourceTypelabelName": sourceTypeLabelName,
+		},
+		templateRegistry...)
+	if err != nil {
+		return nil, err
+	}
+	return &ConfigGenerator{engine}, nil
+}
+
+//Generate the fluent.conf file using the forwarding information
+func (engine *ConfigGenerator) Generate(forwarding *logforward.ForwardingSpec) (string, error) {
+
+	//sanitize here
+	sourceInputLabels := []string{}
+	sourceToPipelineLabels := []string{}
+	pipelineToOutputLabels := []string{}
+	var outputLabels []string
+	var err error
+
+	logTypes := gatherLogSourceTypes(forwarding.Pipelines)
+
+	if sourceInputLabels, err = engine.generateSource(logTypes); err != nil {
+		return "", err
+	}
+	if sourceToPipelineLabels, err = engine.generateSourceToPipelineLabels(mapSourceTypesToPipelineNames(forwarding.Pipelines)); err != nil {
+		return "", err
+	}
+	sort.Strings(sourceToPipelineLabels)
+	if pipelineToOutputLabels, err = engine.generatePipelineToOutputLabels(forwarding.Pipelines); err != nil {
+		return "", err
+	}
+	if outputLabels, err = engine.generateOutputLabelBlocks(forwarding.Outputs); err != nil {
+		return "", err
+	}
+
+	data := struct {
+		CollectInfraLogs        bool
+		CollectAppLogs          bool
+		SourceInputLabels       []string
+		SourceToPipelineLabels  []string
+		PipelinesToOutputLabels []string
+		OutputLabels            []string
+	}{
```

Comment:
this doesn't require a `,` ?

Reply:
> can we define an actual object type for this instead?

Is there a particular reason we need an actual object type instead of anonymous one?